src/WinRT.Runtime/Marshalers.cs

```diff
@@ -25,55 +26,105 @@ public static void Dispose(this GCHandle handle)
                 handle.Free();
             }
         }
+
+#if !NET
+        public static unsafe ref readonly char GetPinnableReference(this string str)
+        {
+            return ref *(char*)((*(byte**)Unsafe.AsPointer(ref str)) + RuntimeHelpers.OffsetToStringData);
+        }
+#endif
     }
 
-    // TODO: minimize heap allocations for marshalers by eliminating explicit try/finally
-    // and adopting ref structs with non-IDisposable Dispose and 'using var ...' pattern,
-    // as well as passing marshalers to FromAbi by ref so they can be conditionally disposed.
     public class MarshalString
     {
-        private IntPtr _header;
-        public GCHandle _gchandle;
-        public IntPtr _handle;
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct HSTRING_HEADER
+        {
+            IntPtr reserved1;
+            int reserved2;
+            int reserved3;
+            int reserved4;
+            int reserved5;
+        }
+
+        private HSTRING_HEADER _header;
+        private GCHandle _gchandle;
+
+        public struct Pinnable

```

Comment:
Should we update our uses of MarshalString in WinRT.Runtime to use the new pinnable behavior where possible?

Reply:
doh!  yes, of course we should (meant to do that)